tests/testthat/test-expand.R

```diff
@@ -44,39 +44,67 @@ test_that("expand accepts expressions", {
   expect_equal(df, crossing(x = 1:3, y = 3:1))
 })
 
-test_that("expand respects groups", {
-  local_options(lifecycle_verbosity = "quiet")
-
+test_that("expand will expand within each group (#396)", {
   df <- tibble(
+    g = c("a", "b", "a"),
     a = c(1L, 1L, 2L),
-    b = c(1L, 2L, 1L),
-    c = c(2L, 1L, 1L)
+    b = factor(c("a", "a", "b"), levels = c("a", "b", "c"))
   )
-  out <- df %>% dplyr::group_by(a) %>% expand(b, c) %>% nest()
+  gdf <- dplyr::group_by(df, g)
+
+  out <- expand(gdf, a, b)
+
+  # Still grouped
+  expect_identical(dplyr::group_vars(out), "g")
+
+  out <- nest(out, data = -g)
 
-  expect_equal(out$data[[1]], crossing(b = 1:2, c = 1:2))
-  expect_equal(out$data[[2]], tibble(b = 1L, c = 1L))
+  expect_identical(out$data[[1]], crossing(a = 1:2, b = factor(levels = c("a", "b", "c"))))
+  expect_identical(out$data[[2]], crossing(a = 1L, b = factor(levels = c("a", "b", "c"))))
 })
 
-test_that("expand allows expanding on grouping variable (#396)", {
+test_that("expand does not allow expansion on grouping variable (#1299)", {
   df <- tibble(
-    g = factor(c("a", "b", "b")),
-    a = c(1, 1, 3)
+    g = "x",
+    a = 1L
   )
   gdf <- dplyr::group_by(df, g)
 
-  out <- expand(gdf, g, a)
+  # This is a dplyr error that we don't own
+  expect_error(expand(gdf, g))
+})
+
+test_that("can use `.drop = FALSE` with expand (#1299)", {
+  levels <- c("a", "b", "c")
 
-  # Same as split by group, expand, combine.
-  # Note that this produces duplicate rows in the result. Not ideal, but
-  # would probably break revdeps if we removed this grouped-df method.
-  expect <- vec_rbind(
-    expand(df[1,], g, a),
-    expand(df[2:3,], g, a)
```

Comment:
I also really like this change because it means that we can no longer produce duplicate rows in the result of `expand()` (which I previously mentioned should be an invariant of this function). This was only an issue if we allow expansion on a grouping column